### PR TITLE
[LibOS] Support C++ programs in SHIM test directories

### PR DESCRIPTION
**Pull request by chiache**
## Affected components

- [ ] README and global configuration
- [ ] Linux PAL
- [ ] SGX PAL
- [ ] FreeBSD PAL
- [ ] Common PAL code
- [x] Library OS (i.e., SHIM), including GLIBC

## Description of the changes 

Fix the Makefile and manifest template in `LibOS/shim/test/native` and `/shim/test/regression` for C++ program. 

## How to test this PR? 

Run the test program:

    cd LibOS/shim/test/regression
    ./pal_loader bootstrap-c++` or `./pal_loader SGX bootstreap-c++`

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/801)

### CODE REVIEW

**Comment by dimakuv** [comment]
Retest Jenkins-SGX please

**Comment by Kushg02** [comment]
Hi all, wanted to ask if the new C++ support for the SHIM test directory programs are related/would resolve an issue I am having running a C++ program I have. 

I am attempting to run a simple program:
```
// C++ program for testing COUT
#include <iostream> 
#include <cstdio> 
//#include <chrono>
//#include <cinttypes>
using namespace std; 

// Driver program to test above functions 
int main() 
{ 
  printf("Hello, World!");
  cout << " seconds";
  return 0; 
}
``` 
However, when I run this segment of code nothing is printed. After enabling inline debugging I found it is a "memory fault at 0x00000000 (IP = 0x3b77487a)". Here is my debug stream: 
```bash
kushgarg@kushgarg-desktop:~/graphene/LibOS/shim/test/apps/gcc$ SGX=1 ./prime.manifest
[P00451] set tcb to 0x3ffe9550
[P00451] host: Linux-SGX
[P00451] bkeep_unmapped: 0x3b6d4000-0x3b6d6000 => vma
[P00451] allocate 0x3b6d4000-0x3b6d6000 for vmas
[P00451] Converting early VMA [0x3bd71060] 0x400000-0x602000
[P00451] Converting early VMA [0x3bd710a8] 0x3ffff000-0x3ffffcdd
[P00451] Converting early VMA [0x3bd710f0] 0x3ba59000-0x3bd75000
[P00451] Converting early VMA [0x3bd71138] 0x3b6d4000-0x3b6d6000
[P00451] heap top adjusted to 0x14eb1000
[P00451] bkeep_unmapped: 0x3b6c3000-0x3b6d4000 => slab
[P00451] bkeep_unmapped: 0x3b6c0000-0x3b6c3000 => slab
[P00451] bkeep_unmapped: 0x3b6bc000-0x3b6c0000 => slab
[P00451] bkeep_unmapped: 0x3b6b7000-0x3b6bc000 => slab
[P00451] bkeep_unmapped: 0x3b6b0000-0x3b6b7000 => slab
[P00451] shim loaded at 0x3ba59000, ready to initialize
[P00451] bkeep_unmapped: 0x3b6af000-0x3b6b0000 => slab
[P00451] bkeep_unmapped: 0x3b6ad000-0x3b6af000 => slab
[P00451] bkeep_unmapped: 0x3b6aa000-0x3b6ad000 => slab
[P00451] mounting default root filesystem
[P00451] bkeep_unmapped: 0x3b6a8000-0x3b6aa000 => slab
[P00451] bkeep_unmapped: 0x3b6a6000-0x3b6a8000 => slab
[P00451] bkeep_munmap: 0x3b6a8000-0x3b6aa000
[P00451] bkeep_munmap: 0x3b6a6000-0x3b6a8000
[P00451] mounting as proc filesystem: /proc
[P00451] mounting as dev filesystem: /dev
[P00451] mounting as chroot filesystem: from dev:tty to /dev
[P00451] bkeep_unmapped: 0x3b6a9000-0x3b6aa000 => slab
[P00451] bkeep_unmapped: 0x3b6a8000-0x3b6a9000 => slab
[P00451] adding port (handle 0x3bfa7d64) for process 0 (type 0001)
[P00451] creating pipe: pipe.srv:1696069657
[P00451] adding port (handle 0x3bfa84e4) for process 0 (type 0002)
[P00451] walk_thread_list(callback=0x3ba99eb2)
[P00451] allocated pid: 1
[P00451] bkeep_unmapped: 0x3b6a3000-0x3b6a8000 => slab
[     1] mounting as chroot filesystem: from file:../../../../../Runtime to /lib
[     1] bkeep_unmapped: 0x3b6a1000-0x3b6a3000 => slab
[     1] bkeep_unmapped: 0x3b69f000-0x3b6a1000 => slab
[     1] bkeep_munmap: 0x3b6a1000-0x3b6a3000
[     1] bkeep_munmap: 0x3b69f000-0x3b6a1000
[     1] bkeep_unmapped: 0x3b6a1000-0x3b6a3000 => slab
[     1] bkeep_unmapped: 0x3b69f000-0x3b6a1000 => slab
[     1] bkeep_munmap: 0x3b6a1000-0x3b6a3000
[     1] bkeep_munmap: 0x3b69f000-0x3b6a1000
[     1] mounting as chroot filesystem: from file:/lib/x86_64-linux-gnu to /lib/x86_64-linux-gnu
[     1] bkeep_unmapped: 0x3b6a1000-0x3b6a3000 => slab
[     1] bkeep_unmapped: 0x3b69f000-0x3b6a1000 => slab
[     1] bkeep_unmapped: 0x3b69c000-0x3b69f000 => slab
[     1] bkeep_munmap: 0x3b69f000-0x3b6a1000
[     1] bkeep_unmapped: 0x3b697000-0x3b69c000 => slab
[     1] bkeep_munmap: 0x3b69c000-0x3b69f000
[     1] bkeep_munmap: 0x3b6a1000-0x3b6a3000
[     1] bkeep_munmap: 0x3b697000-0x3b69c000
[     1] bkeep_unmapped: 0x3b6a1000-0x3b6a3000 => slab
[     1] bkeep_unmapped: 0x3b69f000-0x3b6a1000 => slab
[     1] bkeep_unmapped: 0x3b69c000-0x3b69f000 => slab
[     1] bkeep_munmap: 0x3b69f000-0x3b6a1000
[     1] bkeep_unmapped: 0x3b697000-0x3b69c000 => slab
[     1] bkeep_munmap: 0x3b69c000-0x3b69f000
[     1] bkeep_munmap: 0x3b6a1000-0x3b6a3000
[     1] bkeep_munmap: 0x3b697000-0x3b69c000
[     1] mounting as chroot filesystem: from file:/usr to /usr
[     1] bkeep_unmapped: 0x3b6a1000-0x3b6a3000 => slab
[     1] bkeep_unmapped: 0x3b69f000-0x3b6a1000 => slab
[     1] bkeep_munmap: 0x3b6a1000-0x3b6a3000
[     1] bkeep_munmap: 0x3b69f000-0x3b6a1000
[     1] bkeep_unmapped: 0x3b6a1000-0x3b6a3000 => slab
[     1] bkeep_unmapped: 0x3b69f000-0x3b6a1000 => slab
[     1] bkeep_munmap: 0x3b6a1000-0x3b6a3000
[     1] bkeep_munmap: 0x3b69f000-0x3b6a1000
[     1] mounting as chroot filesystem: from file:/bin to /bin
[     1] bkeep_unmapped: 0x3b6a1000-0x3b6a3000 => slab
[     1] bkeep_unmapped: 0x3b69f000-0x3b6a1000 => slab
[     1] bkeep_unmapped: 0x3b69c000-0x3b69f000 => slab
[     1] bkeep_munmap: 0x3b69f000-0x3b6a1000
[     1] bkeep_munmap: 0x3b6a1000-0x3b6a3000
[     1] bkeep_munmap: 0x3b69c000-0x3b69f000
[     1] bkeep_unmapped: 0x3b6a1000-0x3b6a3000 => slab
[     1] bkeep_unmapped: 0x3b69f000-0x3b6a1000 => slab
[     1] bkeep_unmapped: 0x3b69c000-0x3b69f000 => slab
[     1] bkeep_munmap: 0x3b69f000-0x3b6a1000
[     1] bkeep_munmap: 0x3b6a1000-0x3b6a3000
[     1] bkeep_munmap: 0x3b69c000-0x3b69f000
[     1] bkeep_unmapped: 0x14ab0000-0x14eb1000 => stack
[     1] heap top adjusted to 0x14ab0000
[     1] bkeep_mprotect: 0x14ab1000-0x14eb1000
[     1] allocated stack at 0x14ab1000 (size = 4194304)
[     1] adding file:prime as runtime object loaded at 0x400000-0x602000
[     1] bkeep_mmap: 0x400000-0x401000
[     1] bkeep_mmap: 0x600000-0x602000
[     1] bkeep_mmap: 0x1c0c000-0x3c0c000
[     1] brk area: 0x1c0c000 - 0x1c0d000
[     1] brk reserved area: 0x1c0d000 - 0x3c0c000
[     1] bkeep_mmap: 0x1c0c000-0x1c0d000
[     1] search interpreter: /lib/ld-linux-x86-64.so.2
[     1] bkeep_unmapped: 0x3b69f000-0x3b6a3000 => filebuf
[     1] bkeep_unmapped: 0x1488d000-0x14ab0000
[     1] heap top adjusted to 0x1488d000
[     1] bkeep_mprotect: 0x148ae000-0x14aad000
[     1] bkeep_mmap: 0x14aad000-0x14aaf000
[     1] bkeep_mmap: 0x14aaf000-0x14ab0000
[     1] shim process initialized
[     1] ---- shim_brk(0x0) = 0x01c0c000
[     1] ---- shim_uname(0x14eafb9a) = 0
[     1] ---- shim_mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[P00451] set tcb to 0x3ffd9ee0
[     1] bkeep_unmapped: 0x1488c000-0x1488d000
[P00451] ipc helper thread started
[     1] heap top adjusted to 0x1488c000
[P00451] bkeep_unmapped: 0x3b69a000-0x3b69f000 => slab
[     1] ---- return from shim_mmap(...) = 0x1488c000
[P00451] bkeep_mprotect: 0x3b69b000-0x3b69f000
[P00451] allocated stack at 0x3b69b000 (size = 16384)
[     1] ---- shim_access("/etc/ld.so.preload",F_OK|R_OK) = -2
[P00451] listen to process 0 on port 0x3b6a80c0 (handle 0x3bfa84e4, type 0002)
[     1] ---- shim_open("/lib/tls/libstdc++.so.6",O_RDONLY|2000000,2452570560) ... open
[P00451] listen to process 0 on port 0x3b6a8048 (handle 0x3bfa7d64, type 0001)
[     1] ---- return from shim_open(...) = -2
[     1] ---- shim_stat("/lib/tls",0x14eaf5f0) = -2
[     1] ---- shim_open("/lib/libstdc++.so.6",O_RDONLY|2000000,2452570560) ... open
[     1] ---- return from shim_open(...) = -2
[     1] ---- shim_stat("/lib",0x14eaf5f0) = 0
[     1] ---- shim_open("/lib/x86_64-linux-gnu/tls/libstdc++.so.6",O_RDONLY|2000000,2452570560) ... open
[     1] ---- return from shim_open(...) = -2
[     1] ---- shim_stat("/lib/x86_64-linux-gnu/tls",0x14eaf5f0) = -2
[     1] ---- shim_open("/lib/x86_64-linux-gnu/libstdc++.so.6",O_RDONLY|2000000,2452570560) ... open
[     1] bkeep_unmapped: 0x3b698000-0x3b69a000 => slab
[     1] ---- return from shim_open(...) = -2
[     1] ---- shim_stat("/lib/x86_64-linux-gnu",0x14eaf5f0) = 0
[     1] ---- shim_open("/usr/lib/tls/libstdc++.so.6",O_RDONLY|2000000,2452570560) ... open
[     1] bkeep_unmapped: 0x3b696000-0x3b698000 => slab
[     1] bkeep_unmapped: 0x3b694000-0x3b696000 => slab
[     1] bkeep_unmapped: 0x3b691000-0x3b694000 => slab
[     1] bkeep_munmap: 0x3b694000-0x3b696000
[     1] bkeep_munmap: 0x3b696000-0x3b698000
[     1] bkeep_munmap: 0x3b691000-0x3b694000
[     1] ---- return from shim_open(...) = -2
[     1] ---- shim_stat("/usr/lib/tls",0x14eaf5f0) = -2
[     1] ---- shim_open("/usr/lib/libstdc++.so.6",O_RDONLY|2000000,2452570560) ... open
[     1] ---- return from shim_open(...) = -2
[     1] ---- shim_stat("/usr/lib",0x14eaf5f0) = 0
[     1] ---- shim_open("/usr/lib/x86_64-linux-gnu/tls/libstdc++.so.6",O_RDONLY|2000000,2452570560) ... open
[     1] bkeep_unmapped: 0x3b696000-0x3b698000 => slab
[     1] bkeep_unmapped: 0x3b694000-0x3b696000 => slab
[     1] bkeep_unmapped: 0x3b691000-0x3b694000 => slab
[     1] bkeep_munmap: 0x3b694000-0x3b696000
[     1] bkeep_unmapped: 0x3b68c000-0x3b691000 => slab
[     1] bkeep_munmap: 0x3b691000-0x3b694000
[     1] bkeep_munmap: 0x3b696000-0x3b698000
[     1] bkeep_munmap: 0x3b68c000-0x3b691000
[     1] ---- return from shim_open(...) = -2
[     1] ---- shim_stat("/usr/lib/x86_64-linux-gnu/tls",0x14eaf5f0) = -2
[     1] ---- shim_open("/usr/lib/x86_64-linux-gnu/libstdc++.so.6",O_RDONLY|2000000,2452570560) ... open
[     1] ---- return from shim_open(...) = 3
[     1] ---- shim_read(3,0x14eaf740,832) ... read
[     1] bkeep_unmapped: 0x3b694000-0x3b698000 => filebuf
[     1] ---- return from shim_read(...) = 832
[     1] ---- shim_fstat(3,0x14eaf5f0) = 0
[     1] ---- shim_mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     1] bkeep_unmapped: 0x1488b000-0x1488c000
[     1] heap top adjusted to 0x1488b000
[     1] ---- return from shim_mmap(...) = 0x1488b000
[     1] ---- shim_mmap(0x0,3675136,PROT_READ|PROT_EXEC,MAP_PRIVATE|0x800,3,0) ... mmap
[     1] bkeep_unmapped: 0x14509000-0x1488b000
[     1] heap top adjusted to 0x14509000
[     1] ---- return from shim_mmap(...) = 0x14509000
[     1] ---- shim_mprotect(0x1467b000,2097152,PROT_NONE) ... mprotect
[     1] bkeep_mprotect: 0x1467b000-0x1487b000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mmap(0x1487b000,49152,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|0x800,3,1515520) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x1487b000 with length 49152
[     1] bkeep_mmap: 0x1487b000-0x14887000
[     1] ---- return from shim_mmap(...) = 0x1487b000
[     1] ---- shim_mmap(0x14887000,13312,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON|MAP_FIXED,-1,0) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x14887000 with length 16384
[     1] bkeep_mmap: 0x14887000-0x1488b000
[     1] ---- return from shim_mmap(...) = 0x14887000
[     1] bkeep_munmap: 0x3b694000-0x3b698000
[     1] ---- shim_close(3) = 0
[     1] glibc register library /usr/lib/x86_64-linux-gnu/libstdc++.so.6 loaded at 0x14509000
[     1] ---- shim_open("/lib/libc.so.6",O_RDONLY|2000000,2452570560) ... open
[     1] ---- return from shim_open(...) = 3
[     1] ---- shim_read(3,0x14eaf710,832) ... read
[     1] bkeep_unmapped: 0x3b694000-0x3b698000 => filebuf
[     1] ---- return from shim_read(...) = 832
[     1] ---- shim_fstat(3,0x14eaf5c0) = 0
[     1] ---- shim_mmap(0x0,3852760,PROT_READ|PROT_EXEC,MAP_PRIVATE|0x800,3,0) ... mmap
[     1] bkeep_unmapped: 0x1415c000-0x14509000
[     1] heap top adjusted to 0x1415c000
[     1] ---- return from shim_mmap(...) = 0x1415c000
[     1] ---- shim_mprotect(0x14300000,2093056,PROT_NONE) ... mprotect
[     1] bkeep_mprotect: 0x14300000-0x144ff000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mmap(0x144ff000,24576,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|0x800,3,1716224) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x144ff000 with length 24576
[     1] bkeep_mmap: 0x144ff000-0x14505000
[     1] ---- return from shim_mmap(...) = 0x144ff000
[     1] ---- shim_mmap(0x14505000,14808,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON|MAP_FIXED,-1,0) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x14505000 with length 16384
[     1] bkeep_mmap: 0x14505000-0x14509000
[     1] ---- return from shim_mmap(...) = 0x14505000
[     1] bkeep_munmap: 0x3b694000-0x3b698000
[     1] ---- shim_close(3) = 0
[     1] glibc register library /lib/libc.so.6 loaded at 0x1415c000
[     1] ---- shim_open("/lib/libm.so.6",O_RDONLY|2000000,2442130000) ... open
[     1] ---- return from shim_open(...) = 3
[     1] ---- shim_read(3,0x14eaf5c0,832) ... read
[     1] bkeep_unmapped: 0x3b694000-0x3b698000 => filebuf
[     1] ---- return from shim_read(...) = 832
[     1] ---- shim_fstat(3,0x14eaf470) = 0
[     1] ---- shim_mmap(0x0,3154184,PROT_READ|PROT_EXEC,MAP_PRIVATE|0x800,3,0) ... mmap
[     1] bkeep_unmapped: 0x13e59000-0x1415c000
[     1] heap top adjusted to 0x13e59000
[     1] ---- return from shim_mmap(...) = 0x13e59000
[     1] ---- shim_mprotect(0x13f5b000,2093056,PROT_NONE) ... mprotect
[     1] bkeep_mprotect: 0x13f5b000-0x1415a000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mmap(0x1415a000,8192,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|0x800,3,1052672) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x1415a000 with length 8192
[     1] bkeep_mmap: 0x1415a000-0x1415c000
[     1] ---- return from shim_mmap(...) = 0x1415a000
[     1] bkeep_munmap: 0x3b694000-0x3b698000
[     1] ---- shim_close(3) = 0
[     1] glibc register library /lib/libm.so.6 loaded at 0x13e59000
[     1] ---- shim_open("/lib/libgcc_s.so.1",O_RDONLY|2000000,2442130000) ... open
[     1] ---- return from shim_open(...) = -2
[     1] ---- shim_open("/lib/x86_64-linux-gnu/libgcc_s.so.1",O_RDONLY|2000000,2442130000) ... open
[     1] ---- return from shim_open(...) = 3
[     1] ---- shim_read(3,0x14eaf560,832) ... read
[     1] bkeep_unmapped: 0x3b694000-0x3b698000 => filebuf
[     1] ---- return from shim_read(...) = 832
[     1] ---- shim_fstat(3,0x14eaf410) = 0
[     1] ---- shim_mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     1] bkeep_unmapped: 0x13e58000-0x13e59000
[     1] heap top adjusted to 0x13e58000
[     1] ---- return from shim_mmap(...) = 0x13e58000
[     1] ---- shim_mmap(0x0,2185488,PROT_READ|PROT_EXEC,MAP_PRIVATE|0x800,3,0) ... mmap
[     1] bkeep_unmapped: 0x13c42000-0x13e58000
[     1] heap top adjusted to 0x13c42000
[     1] ---- return from shim_mmap(...) = 0x13c42000
[     1] ---- shim_mprotect(0x13c58000,2093056,PROT_NONE) ... mprotect
[     1] bkeep_mprotect: 0x13c58000-0x13e57000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mmap(0x13e57000,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_FIXED|0x800,3,86016) ... mmap
[     1] mmap: allowing overlapping MAP_FIXED allocation at 0x13e57000 with length 4096
[     1] bkeep_mmap: 0x13e57000-0x13e58000
[     1] ---- return from shim_mmap(...) = 0x13e57000
[     1] bkeep_munmap: 0x3b694000-0x3b698000
[     1] ---- shim_close(3) = 0
[     1] glibc register library /lib/x86_64-linux-gnu/libgcc_s.so.1 loaded at 0x13c42000
[     1] ---- shim_mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     1] bkeep_unmapped: 0x13c41000-0x13c42000
[     1] heap top adjusted to 0x13c41000
[     1] ---- return from shim_mmap(...) = 0x13c41000
[     1] ---- shim_mmap(0x0,8192,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     1] bkeep_unmapped: 0x13c3f000-0x13c41000
[     1] heap top adjusted to 0x13c3f000
[     1] ---- return from shim_mmap(...) = 0x13c3f000
[     1] set tcb to 0x13c3f740
[     1] ---- shim_arch_prctl(4098,0x13c3f740) = 0x00000000
[     1] ---- shim_mprotect(0x144ff000,16384,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x144ff000-0x14503000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mprotect(0x1415a000,4096,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x1415a000-0x1415b000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mmap(0x0,4096,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     1] bkeep_unmapped: 0x13c3e000-0x13c3f000
[     1] heap top adjusted to 0x13c3e000
[     1] ---- return from shim_mmap(...) = 0x13c3e000
[     1] ---- shim_mprotect(0x1487b000,40960,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x1487b000-0x14885000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mprotect(0x600000,4096,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x600000-0x601000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_mprotect(0x14aad000,4096,PROT_READ) ... mprotect
[     1] bkeep_mprotect: 0x14aad000-0x14aae000
[     1] ---- return from shim_mprotect(...) = 0
[     1] ---- shim_brk(0x0) = 0x01c0c000
[     1] brk area: 0x1c0c000 - 0x1c3e000
[     1] brk reserved area: 0x1c3e000 - 0x3c0c000
[     1] bkeep_mmap: 0x1c0c000-0x1c3e000
[     1] ---- shim_brk(0x1c3e000) = 0x01c3e000
[     1] memory fault at 0x00000000 (IP = 0x3b77487a)
[     1] SIGSEGV handled
[     1] run signal handler 0x3ba785b6 (11, 0x14eaf5f0, 0x14eaf678)
[     1] killed by SIGSEGV
[     1] ipc broadcast: IPC_CLD_EXIT(1, 1, 0)
[     1] found port 0x3b6a80c0 (handle 0x3bfa84e4) for process 0 (type 0002)
[     1] found port 0x3b6a8048 (handle 0x3bfa7d64) for process 0 (type 0001)
[     1] parent not here, need to tell another process
[     1] ipc broadcast: IPC_CLD_EXIT(1, 1, 0)
[     1] found port 0x3b6a80c0 (handle 0x3bfa84e4) for process 0 (type 0002)
[     1] found port 0x3b6a8048 (handle 0x3bfa7d64) for process 0 (type 0001)
[     1] this is the only thread 1
[     1] exiting ipc helper
[P00451] ipc helper thread terminated
[     1] deleting port 0x3b6a80c0 (handle 0x3bfa84e4) for process 0
[     1] deleting port 0x3b6a8048 (handle 0x3bfa7d64) for process 0
[     1] process 451 exited with status 0
```
This program functions correctly when running just as an executable: 
```bash
kushgarg@kushgarg-desktop:~/graphene/LibOS/shim/test/apps/gcc$ ./prime
Hello, World! seconds
```

This program also functions correctly when the following program is run (without C++ functions)
```
// C++ program for Quick Sort on Singly Linked List 
//#include <iostream> 
#include <cstdio> 
//#include <chrono>
//#include <cinttypes>
using namespace std; 

// Driver program to test above functions 
int main() 
{ 
	printf("Hello, World!");
  //cout << " seconds";

  //getchar();

  return 0; 
}
```
My manifest appears to be correct in adding the shared libraries from C++ to the manifest template as trusted. I also tried increasing the enclave size to no avail. I dont understand how it could be a memory issue as I have more than 20 GB of availible phyiscal memory. I suspect that it is a C++ support issue one that would likely be fixed in this patch. Would this be a correct assumption?

**Comment by dimakuv** [comment]
Hi @Kushg02 

(I purged your comment a bit. Please do not create such long comments. Also, please open new issues instead of asking on a PR thread.)

When I used this PR, I find no issues running your C++ code. It worked perfectly in all cases. I guess you are right that this PR fixes your particular issue (my guess is that you created your C++ test app in-tree so it used the C compiler with C flags incorrectly, instead of C++ compiler with C++ flags). Another guess is that your manifest was slightly wrong -- though this should have resulted in some intelligible debug message.

Also, the error message you saw has nothing to do with enclave size. The error says `memory fault at 0x00000000` which is a kind of segmentation fault -- your C++/Graphene program tried to dereference a NULL pointer which of course crashes the program. Why there was a dereference of a NULL pointer in the first place, I'm not sure -- as mentioned above, maybe has to do with your C++ program built via a C compiler.

TLDR: You case either (1) use your PR and base your C++ programs on the supplied C++ manifest, or (2) wait until this PR is merged in the master branch (~2 weeks from now).
